### DISCUSSION TITLE
Linux : add a symbolic link in an ANE

### DISCUSSION THREAD

**Discussion by pol2095** [airsdk/Adobe-Runtime-Support#1803] [Q&A] [answered]
Hello,

I tried to add a symbolic link in a ANE, it's a symbolic link to a myLib.so.4.1.2 shared library
using ADT it doesn't work, the symbolic link is replaced by the shared library file without version (myLib.so)

I tried too to add the symbolic link in the ANE like a zip : 
```zip --symlinks -ur myAne.ane META-INF/```
but later it's impossible to compile the captive app using the ANE : `Packaging failed. Packager internal error`

Thanks

**Comment by ajwfrost** [comment]
Hi

I'm not quite sure what you're trying to do here? Do you want to have an ANE that has just a symbolic link to a library that's not included in the ANE itself? I'm not quite sure why you'd want that but the risk would then be that it's installed onto a computer without the library that it links to?

If your ANE has a dependency on a library that's external then you probably don't need to include the link, assuming the library is expected to be present in the normal library folder of the target machine. Or you could write a shim library that is packaged like normal into the ANE but just searches for and opens up the real target library and then passes through the API calls?

thanks

**Comment by pol2095** [comment]
CMake generate on Ubuntu `myLib.so.4.1.0` and a symbolic link `myLib.so`
for working my ANE need `myLib.so` and `myLib.so.4.1.0`
I have a duplicate file in the ANE

I thinked the ANE is a zip file, `zip --symlinks -ur myAne.ane META-INF/` create a valid zip but the ANE doesn't work
how `adt` create the ANE, is-it signed ?

**Comment by ajwfrost** [comment]
Ah I see, so you are generating the library file with version information (`myLib.so.4.1.0`) and then have a symbolic link to this (`myLib.so`) and are then using this in the ANE either via `dlopen("myLib.so")` or by specifying `myLib.so` as the native library...

It doesn't appear that we support symbolic links within the ANE itself (as you've found .. yes an ANE is a zip file but it's processed by our Java ADT toolchain so needs to conform to the support that we have in that). The whole thing would work perfectly fine if you renamed `myLib.so.4.1.0` to `myLib.so` and used it as the main binary rather than as the link, so we can perhaps update ADT's ANE packaging mechanism to do that automatically... you may end up then with a duplicate if you include both libraries but if you just included the `myLib.so` link file, we can follow that and package up the actual library instead...

thanks

**Comment by pol2095** [comment]
I tried too to rename `myLib.so.4.1.0` in `myLib.so` but it doesn't work
for working it's necessary to have a file `myLib.so` and a file `myLib.so.4.1.0`

**Comment by pol2095** [comment]
I modified "CMakeLists.txt", I commented lines with properties [VERSION](https://cmake.org/cmake/help/latest/prop_tgt/VERSION.html#prop_tgt:VERSION) and [SOVERSION](https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html#prop_tgt:SOVERSION)
It have just a file `myLib.so` now, thanks.

**Comment by pol2095** [comment]
Hello, the ANE is correctly packaged with the symlink, but `adt -package -target bundle ... -extdir libs` replace the symlink by a file `myLib.so.4.1.0` with the content "myLib.so" ?

**Reply by ajwfrost** [reply]
I do recall we did some changes to allow symbolic links in more than just the macOS Frameworks which is how it had originally been set up. But the packaging sounds like it's then a problem and is getting the Windows-style result of a symbolic link being just a file containing the relative path/filename that it links to.

Are you able to share the ANE with us so that we can reproduce and check it? Let me know if you need a link for uploads..

thanks

**Reply by pol2095** [reply]
You can download a test ANE [here](https://pol2095.eu/Air/TestSymlink/Linux_TestSymlinkv0.0.1.zip)

**Reply by ajwfrost** [reply]
Thanks for that. So it looks like the native code that assembles the bundle is just using the normal Java zip classes - which don't support symbolic links. The packaging code for ANEs/AIR files is lower level where we're writing into a random-access file and creating the zip file by generating the local file headers and central directory records etc.

Options then are:
- write our own zip decompression class to match the writing of the zip files. It should only be handling files that it's created which may mean this isn't too much work..
- integrate a third party solution such as [Apache Commons Compress](https://commons.apache.org/proper/commons-compress/zip.html) 

No quick-fix I'm afraid, unless we did a kind of hack solution where we look at a ZipEntry where we know it's part of a Linux ANE that's being added to a bundle, and see whether the contents of this file match another ZipEntry filename... but that would be a major hack and potentially could break things in case someone had a valid reason for having this scenario.

thanks

**Reply by pol2095** [reply]
I agree that integrating [Apache Commons Compress](https://commons.apache.org/proper/commons-compress/zip.html) is the best approach. You could use it to detect the symlink entries in the archive, and then simply rely on java.nio.file.Files.createSymbolicLink() to generate them on the OS. Thanks!

**Reply by ajwfrost** [reply]
Actually .. could you please check whether it works with the attached (adt.jar within the zip file here)
[adt_1803.zip](https://github.com/user-attachments/files/28171789/adt_1803.zip)

**Reply by pol2095** [reply]
The updated `adt.jar` in `<AIR_SDK>/lib` doesn't fix the `adt` problem :
<img width="346" height="152" alt="Capture d’écran du 2026-05-23 09-55-46" src="https://github.com/user-attachments/assets/9c1c3413-38db-4826-9909-cc1efc3db290" />

with `adl`, I unzip manually the ANE :
<img width="346" height="152" alt="Capture d’écran du 2026-05-23 09-57-34" src="https://github.com/user-attachments/assets/126d31aa-abc2-4eff-9dc6-3c0f196a575e" />

**Reply by ajwfrost** [reply]
Hi @pol2095  .. it turns out, there's a lot more to it than I had thought. It's not ADT that needed support - this already had it, Adobe had a simpler class that had been created for use alongside the Java zip classes to be able to support the unix properties for zip entries. But instead, the AIR runtime itself needed to support symbolic links on Linux which (oddly!) it didn't..

One issue was that, if you use `File.copyTo()` to copy a folder structure on Linux, it will follow links and replace them with the actual file. We've adjusted this so that it copies the links as-is i.e. a link stays a link with the same contents - but only for the SDK tools, rather than changing this for everyone, as that's quite a change in behaviour. But one that I think we should expose through an API change at some point, as it could be quite useful for people to affect this. (On macOS, the copy operating is handed off to an OS component in its entirety, so I would assume that this already keeps links as links..)

Anyway -> are you able to check using the below libCore.so file which should be put into your AIR SDK under `runtimes/air/linux-x64/Adobe\ AIR/Versions/1.0/`
https://transfer.harman.com/link/4LPZEVK4DDk3Dtsv5TDe3L

thanks

**Reply by pol2095** [reply]
using the new libCore.so on Linux amd64, I have this message error with `adl` `Error loading the runtime (libssl.so.1.1: cannot open shared object file: No such file or directory)` and with `adt` : `Failed to load the AIR Runtime`

**Reply by ajwfrost** [reply]
Ah - sorry, it was built from a dev machine rather than the proper environment so (a) won't work on older machines, and (b) the linker settings wouldn't have worked to configure it properly. Can you run:
```
export LD_LIBRARY_PATH=/path/to/AIR/SDK/runtimes/air/linux-x64/Adobe\ AIR/Versions/1.0/Resources
```
Then if you run it from that terminal, it should work.

Failing that, I'll kick off a build from the production environment..

thanks

**Reply by pol2095** [reply]
now with the updated `adt.jar,` `libCore.so` and `export LD_LIBRARY_PATH=/path/to/AIR/SDK/runtimes/air/linux-x64/Adobe\ AIR/Versions/1.0/Resources` the symlink is correctly created.

**Reply by ajwfrost** [reply]
Great! Thanks for confirming; we will get these changes into our next release..

many thanks

**Reply by pol2095** [reply]
I tried Air 51.3.3.1, sometimes symlinks are correctly created, sometimes no
<img width="325" height="127" alt="Capture d’écran du 2026-06-16 17-54-19" src="https://github.com/user-attachments/assets/277c3f45-8b9e-485a-b460-8ee8bf2d62d5" />
<img width="325" height="127" alt="Capture d’écran du 2026-06-16 17-52-41" src="https://github.com/user-attachments/assets/b2a2c589-64f3-49f8-a1a2-93d666299ffe" />

**Reply by ajwfrost** [reply]
Hmm .. is there any pattern on whether it works or not, or is this seemingly just random? And can you do an `ls -la` of the resulting files so that we can see what's happened there, is it just again a text file with the contents "libTestSymlink.so.2"?

**Reply by pol2095** [reply]
It isn't a text file, it's a dangling symlink.

`ls -la libTestSymlink.so` :
when it doesn't work
```
lrwxrwxrwx 1 pol2095 pol2095 22 juin  16 18:31 libTestSymlink.so -> 'libTestSymlink.so.2'$'\220\366\024'
```

when it work
```
lrwxrwxrwx 1 pol2095 pol2095 19 juin  16 18:31 libTestSymlink.so -> libTestSymlink.so.2
```

**Reply by ajwfrost** [reply]
Ah ... could be some issue with string terminations... thanks for that! let me check..

**Reply by pol2095** [reply]
Any news ?

**Reply by ajwfrost** [reply]
We've not been able to reproduce this I'm afraid. But, having done a review of the updates and the documentation, we did notice a little comment within the notes for "readlink" which suggests that we should not expect the returned link string to be null-terminated .. so we think this is the most likely culprit.

Can I check if you're using x86_64 or arm64, we can perhaps send you a test build to check this change has worked?

thanks

**Reply by pol2095** [reply]
I tried on Ubuntu amd64 but I can try on a RPi5 arm64 too, send me a test build.

**Reply by ajwfrost** [reply]
ARM64 version for a Pi:
https://transfer.harman.com/message/ZN13EqGVokv8dNhdaEqoSl
thanks

**Reply by pol2095** [reply]
Now that the symlink seems to be created correctly without any extra string terminations, could you send me an x86_64 test build as well ?

**Reply by ajwfrost** [reply, marked as answer]
Sure, please check the later .so file:
https://transfer.harman.com/message/IGuUSEuECnaZkKdvKzAP5u

thanks

**Reply by pol2095** [reply]
Great job, the symlink is created correctly on x86_64 and arm64 (without any extra string terminations).

**Reply by ajwfrost** [reply]
Great, thank you for confirming! Will get that into our next release..
